src/zenml/client_lazy_loader.py

```diff
@@ -0,0 +1,191 @@
+
+        Args:
+            args: Positional arguments.
+            kwargs: Keyword arguments.
+
+        Returns:
+            self
+        """
+        # workaround to protect from infinitely looping over in deepcopy called in invocations
+        if not self._exclude_next_call:
+            self.call_chain.append(
+                _CallStep(is_call=True, call_args=args, call_kwargs=kwargs)
+            )
+        self._exclude_next_call = False
+        return self
+
+    def __getitem__(self, item: Any) -> "ClientLazyLoader":
+        """Get item from mocked attribute.
+
+        Args:
+            item: Item to get.
+
+        Returns:
+            self
+        """
+        self.call_chain.append(_CallStep(selector=item))
+        return self
+
+    def evaluate(self) -> Any:
+        """Evaluate lazy loaded Client method.
+
+        Returns:
+            Evaluated lazy loader chain of calls.
+        """
+        from zenml.client import Client
+
+        def _iterate_over_lazy_chain(
+            self: "ClientLazyLoader", self_: Any, call_chain_: List[_CallStep]
+        ) -> Any:
+            next_step = call_chain_.pop(0)
+            if next_step.is_call:
+                self_ = self_(*next_step.call_args, **next_step.call_kwargs)
+            elif next_step.selector:
+                self_ = self_[next_step.selector]
+            elif next_step.attribute_name:
+                self_ = getattr(self_, next_step.attribute_name)
+            else:
+                raise RuntimeError(
+                    f"Failed to evaluate lazy load chain `{self.method_name}` "
+                    f"+ `{self.call_chain}`. Reach out to the ZenML team via "
+                    "Slack or GitHub to check further."
+                )
+            return self_
+
+        self_ = getattr(Client(), self.method_name)
+        call_chain_ = self.call_chain.copy()
+        while call_chain_:
+            self_ = _iterate_over_lazy_chain(self, self_, call_chain_)
+        return self_
+
+
+def client_lazy_loader(
+    method_name: str, *args: Any, **kwargs: Any
+) -> Optional[ClientLazyLoader]:
+    """Lazy loader for Client methods helper.
+
+    Usage:
+    ```
+    def get_something(self, arg1: Any)->SomeResponse:
+        if cll:=client_lazy_loader("get_something", arg1):
+            return cll # type: ignore[return-value]
+        return SomeResponse()
+    ```
+
+    Args:
+        method_name: The name of the method to be called.
+        *args: The arguments to be passed to the method.
+        **kwargs: The keyword arguments to be passed to the method.
+
+    Returns:
+        The result of the method call.
+    """
+    from zenml import get_pipeline_context
+
+    try:
+        get_pipeline_context()
+        cll = ClientLazyLoader(
+            method_name=method_name,
+        )
+        return cll(*args, **kwargs)
+    except RuntimeError:
+        return None
+
+
+def evaluate_all_lazy_load_args_in_client_methods(
```

Comment:
Can you help me understand what this decorator does? It seems to create a `ClientLazyLoader` for all dict values passed to Client methods, but I'm not sure why this is necessary.

Also, the `ClientLazyLoader` has a required field called `method_name` which I don't see being populated by these dict values?

Reply:
What I don't understand is, shouldn't this (`if isinstance(args_[i], dict):`) be `if isinstance(args_[i], ClientLazyLoader):` instead? Or will the lazy loaders always be converted to a dict in this case?

Reply:
I think it would be good also to check for `ClientLazyLoader` in this case, though in theories the only use case I see is that this method was somehow passed as part of input to one of the steps, so it will be dict after reading it from DB. But it is no harm to extend it to check for class instance itself.

Reply:
Pushed the change